README.md
# audiorecorder

####  A free tool for the calibration and recording of analog audio signals

## Development information:
For updates and history of development please see [this page](https://github.com/amiaopensource/audiorecorder/blob/master/history.md).

## Installation:

__Audiorecorder2 Note:__ Audiorecorder2 can now be installed along side audiorecorder by adding the flag `--with-audiorecorder2` to the following homebrew install commands.

audiorecorder can be installed via [homebrew](https://brew.sh/) on macOS and [linuxbrew](http://linuxbrew.sh/) on Linux.

__MacOS Installation__:
Audiorecorder can be installed via Homebrew with the following commands:

`brew tap amiaopensource/amiaos`

`brew install audiorecorder`

__Linux Installation__ (Tested in Ubuntu 16.04)

The following division of commands seeks to install up-to-date versions of audiorecorder dependencies while avoiding conflicts.

__Standard package manager commands__

`sudo apt-get install sox`

`sudo apt-get install libsdl2-2.0`

Install current [MPV](https://mpv.io/installation/) ppa, then `sudo apt-get install mpv`

__linuxbrew commands__

`brew install dialog`

If FFmpeg has already been installed with these options then skip next step. 

`brew install ffmpeg --with-freetype`

sdl2 is necessary to build FFmpeg with FFplay, but removing it after build will avoid conflicts between other installations of sdl2.

`brew uninstall --ignore-dependencies sdl2`

`brew install --ignore-dependencies audiorecorder`

## Usage:
Usage: [-p] passthrough mode, [-e] edit config, [-m] edit metadata for BWF insertion

#### Initial setup:
To set up audiorecorder, first run the command `audiorecorder -e`. This will open an interface where settings such as recording bit depth and sample rate can be selected. This is also where the option to embed BEXT metadata can be selected.

To set the metadata that will be embedded in the BEXT chunk (if activated in the configuration menu) run the command `audiorecorder -m`. This will open an interface where the desired values can be entered.

#### Explanation of Interface:
1: Peak volume meters (dB)

2: Audio spectrum - lines (from left) represent 1, 5, 10, 15, 20kHz respectively

3: Recording time

4: Spectrogram

5: Audio vecorscope

6: Graph of peak volume values (dB) from -30dB to 0dB. Colored area represents -5dB to 0dB.

<img src="https://raw.githubusercontent.com/amiaopensource/audiorecorder/master/numbered_interface.png" alt="audiorecorder interface" height="350" width="350">

#### Preview:
Running `audiorecorder -p` will open the preview window. This window can be closed with the escape key. This allows you to check your settings and playback machine calibration using the audiorecorder interface without creating a file. It is recommended to use this mode before every recording. Once you are satisfied that everything is calibrated appropriately, move on to record mode.

#### Record:
Running the command `audiorecorder` will start record mode. You will be prompted for the name of the file created. __Wait until the audiorecorder interface opens__, and then press play on your playback machine. During recording do not resize the window, as this can cause problems with the audio buffer. To stop recording press the escape key.

Once recording is complete, audiorecorder will process the file, and then an interface will open showing the waveform of the recorded file, as well as giving options to 'preview' and 'trim' the file.

Press 'Preview' to hear the file you recorded. To trim file, enter the amount (in seconds) to trim from the start and end of the file and press 'Trim'. If 'Start Trim' is left blank, auto-trim will be applied to start of file. If no trim at start is desired enter '0'. After trimming, a preview window will open for your new file. Trim can be run as many times as is necessary. Once you are done, press 'Finish' to finalize and quit. It is at this point that BEXT metadata will be embedded if that option has been selected.

## Huge thanks to all contributors to audiorecorder!

#### Project Maintainers are:
Andrew Weaver (@privatezero)

Ashley Blewer (@ablwr)

#### Contributors to date:
privatezero (Andrew Weaver), retokromer (Reto Kromer), dericed (Dave Rice), todrobbins (Tod Robbins), ablwr (Ashley Blewer)

Special thanks to Matt Boyd at the University of Washington for extensive testing assistance!

## License

The audiorecorder script is covered under a [BSD 3-clause license](https://github.com/amiaopensource/audiorecorder/blob/master/LICENSE.txt)

<a rel="license" href="https://creativecommons.org/licenses/by/4.0/"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by/4.0/88x31.png"></a><br>All documentation for audiorecorder is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution 4.0 International License</a>.


history.md
# Development History

**2018-08-30:** audiorecorder2 added to homebrew install formula. Can now be installed with option `--with-audiorecorder2`

**2018-03-15:** In development version 2 (in linux) tested with twenty minute recording at 24 bit/96 kHz stereo recording. File analyzed for interstitial errors using Wavelab Elements 9.5 global analysis with none detected.

**2017-12-26:** Version 0.1.08 released. Mainly typo fixes.

**2017-12-14:** Version 0.1.07 released. Fix bug. Stricter man page syntax.

**2017-10-18:** Version 0.1.06 released. Improved help message and man page.

**2017-10-01:** Version 0.1.05 released. Deletes BWF Metaedit from the package and uses its homebrew formula instead.

**2017-08-22:** Version 0.1.04 released. Mainly doc fixes.

**2017-06-05:** Version 0.1.03 released. Linux install instructions, man page and Code of Conduct have been added.

**2017-04-25:** Version 0.1.02 released. Includes font support for filters in Linux. (Tested in Ubuntu 16.04).

**2017-04-11:** Version 0.1.01 released. Contains some basic interface workarounds to decrease reliance on Pashua for user input. This makes audiorecorder able to run in a linux environment (further reliability testing for non macOS systems is underway).

**2017-03-11:** Major redesign of post-digitization functions and GUI. Audiorecorder now supports trimming of start and end of files via manual specification as well as auto-trim of silence at the start of files. Preview and post-digitization GUI now incorporate waveforms of the digitized audio to aid in trimming. All testing has been negative for dropped samples.

**2017-03-04:** For the new release changes have been made to streamline the way audiorecorder handles input channels. This will prevent it from being overwhelmed by A/D converters that output many empty tracks along with the desired tracks (such as the Apogee Symphony). In two 55 minute tests and one 30 minute no drops or problems were detected via Wavelab analysis. Due to more efficient management of data the current buffering now has increased the latency for mono captures. A next step will be to see how much buffering can be safely lowered.

**2017-02-24:** Changes appear to have been successful for adapting audiorecorder to macOS 10.12.03. In testing it was discovered that audiorecorder has issues with A/D converters that have a large number of output tracks (such as sixteen as opposed to two). Current audiorecorder head appears to be stable for two track A/D converters and testing is underway to expand its ability to deal with larger multi-track converters.

**2017-02-23:** Due to some issues that were reported with macOS 10.12.03, some changes to buffering have been made and are currently being tested.

**2017-02-15:** Ability to select which channel to record has been added to the GUI. This option was tested with a 25 minute ingest on 2013 Macbook Air with no dropped samples detected audibly or vie Wavelab global analysis. Some glitches with file trimming were resolved.

**2017-01-20:** Initial post digitization functions have been included via a basic GUI. These include ability to preview file with and without silence trimming, and an option to create a silence trimmed version of file. The spectrograph has also been changed to scroll for easier viewability. Further testing has not detected any dropped samples either audibly or via Wavelab global analysis. 

**2016-12-22:** Current build was tested with a 40 minute transfer on the 2013 Macbook Air with no dropped samples detected either audibly or via Wavelab's global analysis tool. Testing continues...


audiorecorder.sh
#!/usr/bin/env bash
#Cascadia Now!
config="$HOME/.$(basename "${0}").conf"
bwf_config="$HOME/.$(basename "${0}")_BWF.conf"
touch "${config}"
touch "${bwf_config}"
. "$config"
. "${bwf_config}"



OPTIND=1
while getopts "epm" opt ; do
    case "${opt}" in
        e) runtype="edit";;
        p) runtype="passthrough";;
        m) runtype="metadata";;
        *)
    esac
done

_lookup_devices(){
    OLDIFS="$IFS"
    device_list=("$(ffmpeg -hide_banner -f avfoundation -list_devices true -i "" -f null /dev/null 2>&1 | cut -d ' ' -f6- | sed -e                  '1,/AVFoundation audio devices:/d' | grep -v '^$'| tr '\n' ',' )")
    IFS=',' read -r -a DEVICES <<< "$device_list"
    IFS="$OLDIFS"
}
_lookup_sample_rate(){
    case "${1}" in
        "44.1 kHz")
            SAMPLE_RATE_NUMERIC="44100"
            ;;
        "48 kHz")
            SAMPLE_RATE_NUMERIC="48000"
            ;;
        "96 kHz")
            SAMPLE_RATE_NUMERIC="96000"
            ;;
    esac
}

_lookup_bit_depth(){
    case "${1}" in
        "16 bit")
            CODEC="pcm_s16le"
            ;;
        "24 bit")
            CODEC="pcm_s24le"
            ;;
    esac
}
_get_ITEM_ID(){
    echo "Please Input Item ID or type Q to exit"
    read ITEM_ID
    if [ "${ITEM_ID}" = "Q" ] || [ "${ITEM_ID}" = "q" ] ; then
        echo "Goodbye!"
        exit 0
    fi
    if [ -f "${output}/${ITEM_ID}.wav" ] ; then
        echo "A file with that name already exists!"
        _get_ITEM_ID
    fi
}

_metadata_gui(){
    gui_conf="
    # Set transparency: 0 is transparent, 1 is opaque
    *.transparency=0.95
    # Set window title
    *.title = Edit Metadata Values
    # intro text
    intro.width = 300
    intro.type = text
    intro.text = Hello
    #Originator
    originator.type = textfield
    originator.label = Enter Originator
    originator.default = "${originator}"
    #Coding History
    coding_history.type = textbox
    coding_history.label = Enter Coding History
    coding_history.default = "${coding_history}"
    cb.type = cancelbutton
    cb.label = Cancel
    "

    pashua_configfile=`/usr/bin/mktemp /tmp/pashua_XXXXXXXXX`
    echo "${gui_conf}" > "${pashua_configfile}"
    pashua_run
    rm "${pashua_configfile}"
}

_master_gui(){
    _lookup_devices    
    gui_conf="
    # Set transparency: 0 is transparent, 1 is opaque
    *.transparency=0.95
    # Set window title
    *.title = Welcome to Audio Recorder!
    # intro text
    intro.width = 300
    intro.type = text
    intro.text = Hello
    #Output Location
    output.type = openbrowser
    output.label = Output Location
    output.default = "${output}"
    #Capture Device
    device.type = popup
    device.label = Select Audio Capture Device
    device.option = ${DEVICES[0]}
    device.option = ${DEVICES[1]}
    device.option = ${DEVICES[2]}
    device.option = ${DEVICES[3]}
    device.default = "${device}"
    #Sample Rate
    sample_rate.type = radiobutton
    sample_rate.label = Select Sample Rate
    sample_rate.option = 44.1 kHz
    sample_rate.option = 48 kHz
    sample_rate.option = 96 kHz
    sample_rate.default = "${sample_rate}"
    #Bit Depth
    bit_depth.type = radiobutton
    bit_depth.label = Select Bit Depth
    bit_depth.option = 16 bit
    bit_depth.option = 24 bit
    bit_depth.default = "${bit_depth}"
    #Cancel Button
    cb.type = cancelbutton
    cb.label = Cancel
    "

    pashua_configfile=`/usr/bin/mktemp /tmp/pashua_XXXXXXXXX`
    echo "${gui_conf}" > "${pashua_configfile}"
    pashua_run
    rm "${pashua_configfile}"
}
pashua_run() {
    # Wrapper function for interfacing to Pashua. Written by Carsten
    # Bluem <carsten@bluem.net> in 10/2003, modified in 12/2003 (including
    # a code snippet contributed by Tor Sigurdsson), 08/2004 and 12/2004.
    # Write config file

    # Find Pashua binary. We do search both . and dirname "$0"
    # , as in a doubleclickable application, cwd is /
    # BTW, all these quotes below are necessary to handle paths
    # containing spaces.
    bundlepath="Pashua.app/Contents/MacOS/Pashua"
    mypath=$(dirname "${0}")
    for searchpath in "$mypath/Pashua" "$mypath/$bundlepath" "./$bundlepath" \
                      "/Applications/$bundlepath" "$HOME/Applications/$bundlepath"
    do
        if [ -f "$searchpath" -a -x "$searchpath" ] ; then
            pashuapath=$searchpath
            break
        fi
    done
    if [ ! "$pashuapath" ] ; then
        echo "Error: Pashua is not found."
                break 2
            fi
        encoding=""
        # Get result
        result=`"$pashuapath" $encoding $pashua_configfile | sed 's/ /;;;/g'`

        # Parse result
        for line in $result ; do
            key=`echo $line | sed 's/^\([^=]*\)=.*$/\1/'`
            value=`echo $line | sed 's/^[^=]*=\(.*\)$/\1/' | sed 's/;;;/ /g'`
            varname=$key
            varvalue="$value"
            eval $varname='$varvalue'
        done
} # pashua_run()

if [ "${runtype}" = "edit" ] ; then
_master_gui

{
    echo "output=\"${output}\""
    echo "device=\"${device}\""
    echo "sample_rate=\"${sample_rate}\""
    echo "bit_depth=\"${bit_depth}\""
} > "${config}"
fi

if [ "${runtype}" = "metadata" ] ; then
_metadata_gui

{
    echo "originator=\"${originator}\""
    echo "coding_history=\"${coding_history}\""
} > "${bwf_config}"
fi

if [ -n ${DEVICE_NUMBER} ] ; then
    _lookup_sample_rate "${sample_rate}"
else 
    echo "No Sample Rate Specified.  Setting to 96kH."
    SAMPLE_RATE_NUMERIC="96000"
fi
if [ -n "${bit_depth}" ] ; then
    _lookup_bit_depth  "${bit_depth}"
else 
    echo "No Bit Depth Specified.  Setting to 24."
    CODEC="pcm_s24le"
fi
if [ -n "${device}" ] ; then
    DEVICE_NUMBER=$(echo "${device}" | cut -c 2)
else 
    echo "No Device Specified.  Attempting to Guess Device."
    DEVICE_NUMBER="0"
fi
if [ -z "${output}" ] ; then
    echo "No Output Location Specified.  Setting Output to Desktop"
    output="~/Desktop"
fi


FILTER_CHAIN="asplit=6[out1][a][b][c][d][e],\
[e]showvolume=w=700:c=0xff0000:r=30[e1],\
[a]showfreqs=mode=bar:cmode=separate:size=300x300:colors=magenta|yellow[a1],\
[a1]drawbox=12:0:3:300:white@0.2[a2],[a2]drawbox=66:0:3:300:white@0.2[a3],[a3]drawbox=135:0:3:300:white@0.2[a4],[a4]drawbox=202:0:3:300:white@0.2[a5],[a5]drawbox=271:0:3:300:white@0.2[aa],\
[b]avectorscope=s=300x300:r=30:zoom=5[b1],\
[b1]drawgrid=x=150:y=150:c=white@0.3[bb],\
[c]showspectrum=s=400x600:mode=combined:color=rainbow:scale=lin:saturation=4[cc],\
[d]astats=metadata=1:reset=1,adrawgraph=lavfi.astats.Overall.Peak_level:max=0:min=-30.0:size=700x256:bg=Black[dd],\
[dd]drawbox=0:0:700:42:hotpink@0.2:t=42[ddd],\
[aa][bb]vstack[aabb],[aabb][cc]hstack[aabbcc],[aabbcc][ddd]vstack[aabbccdd],[e1][aabbccdd]vstack[z],\
[z]drawtext=fontfile=/Library/Fonts/Andale Mono.ttf: text='%{pts \\: hms}':x=460: y=50:fontcolor=white:fontsize=30:box=1:boxcolor=0x00000000@1[fps],[fps]fps=fps=30[out0]"


if [ "${runtype}" = "passthrough" ] ; then
    ffmpeg -f avfoundation -i "none:"${DEVICE_NUMBER}"" -f wav -c:a pcm_s16le -ar 44100 - |\
    ffplay -window_title "Audio Record" -f lavfi \
    "amovie='pipe\:0',${FILTER_CHAIN}"
    exit
fi

_get_ITEM_ID

mkfifo PIPE2REC
ffmpeg -f avfoundation -i "none:"${DEVICE_NUMBER}"" -f wav -c:a "${CODEC}" -ar "${SAMPLE_RATE_NUMERIC}" -y PIPE2REC -f wav -c:a pcm_s16le -ar 44100 - |\
ffplay -window_title "Audio Record" -f lavfi \
"amovie='pipe\:0',${FILTER_CHAIN}" | ffmpeg -i PIPE2REC -c copy -rf64 auto "${output}"/"${ITEM_ID}".wav


# Embed metadata in BEXT
echo "Preparing to embed metadata"
echo ""
if
	(("${#ITEM_ID}" > 32));
then
	orig_ref="See description for Identifiers"
else
    orig_ref="${ITEM_ID}"
fi

bwfmetaedit --reject-overwrite --Description="${ITEM_ID}".wav --Originator="${originator}" --OriginatorReference="${orig_ref}" --History="${coding_history}" --IARL="${originator}" --MD5-Embed --OriginationDate=$(date "+%Y-%m-%d") --OriginationTime=$(date "+%H:%M:%S") "${output}"/"${ITEM_ID}".wav


audiorecorder2-dev/audiorecorder2/audiorecorder2.rb
require 'yaml'
require 'json'

# Recording Variables
if RUBY_PLATFORM.include?('linux')
  Drawfontpath = '/usr/share/fonts/truetype/liberation/LiberationSans-Regular.ttf'
  Soxpath = 'rec'
  Ffmpegpath = 'ffmpeg'
  Ffplaypath = 'ffplay'
  Ffprobepath = 'ffprobe'
  Bwfmetaeditpath = 'bwfmetaedit'
  Mpvpath = 'mpv'
elsif RUBY_PLATFORM.include?('darwin')
  Drawfontpath = '/Library/Fonts/Andale Mono.ttf'
  Soxpath = '/usr/local/bin/rec'
  Ffmpegpath = '/usr/local/bin/ffmpeg'
  Ffplaypath = '/usr/local/bin/ffplay'
  Ffprobepath = '/usr/local/bin/ffprobe'
  Bwfmetaeditpath = '/usr/local/bin/bwfmetaedit'
  Mpvpath = '/usr/local/bin/mpv'
else
  Drawfontpath = 'some windows path'
end

# Check FFmpeg version for filter options

if system('#{Ffmpegpath} -version | grep "ffmpeg version 4.0.0\|ffmpeg version 4.0.1\|ffmpeg version 4.0.2"')
    Avectorscopefilter = 'avectorscope=s=300x300:r=30:zoom=5:mirror=0'
else
    Avectorscopefilter = 'avectorscope=s=300x300:r=30:zoom=5'
end

FILTER_CHAIN = "asplit=6[out1][a][b][c][d][e],\
[e]showvolume=w=700:c=0xff0000:r=30[e1],\
[a]showfreqs=mode=bar:cmode=separate:size=300x300:colors=magenta|yellow[a1],\
[a1]drawbox=12:0:3:300:white@0.2[a2],[a2]drawbox=66:0:3:300:white@0.2[a3],[a3]drawbox=135:0:3:300:white@0.2[a4],[a4]drawbox=202:0:3:300:white@0.2[a5],[a5]drawbox=271:0:3:300:white@0.2[aa],\
[b]#{Avectorscopefilter}[b1],\
[b1]drawgrid=x=150:y=150:c=white@0.3[bb],\
[c]showspectrum=s=400x600:slide=scroll:mode=combined:color=rainbow:scale=lin:saturation=4[cc],\
[d]astats=metadata=1:reset=1,adrawgraph=lavfi.astats.Overall.Peak_level:max=0:min=-30.0:size=700x256:bg=Black[dd],\
[dd]drawbox=0:0:700:42:hotpink@0.2:t=42[ddd],\
[aa][bb]vstack[aabb],[aabb][cc]hstack[aabbcc],[aabbcc][ddd]vstack[aabbccdd],[e1][aabbccdd]vstack[z],\
[z]drawtext=fontfile=#{Drawfontpath}: text='%{pts \\: hms}':x=460: y=50:fontcolor=white:fontsize=30:box=1:boxcolor=0x00000000@1[fps],[fps]fps=fps=30[out0]"

# Set Configuration
sox_channels = '1 2'
ffmpeg_channels = 'stereo'
$codec_choice = 'pcm_s24le'
$soxbuffer = '50000'
$sample_rate_choice = '96000'
$filename = ''

# Load options from config file
configuration_file = File.expand_path('~/.audiorecorder2.conf')
if ! File.exist?(configuration_file)
  config_options = "destination:\nsamplerate:\nchannels:\ncodec:\norig:\nhist:\nbext:"
  File.write(configuration_file, config_options)
end
config = YAML::load_file(configuration_file)
$outputdir = config['destination']
$sample_rate_choice = config['samplerate']
sox_channels = config['channels']
$codec_choice = config['codec']
$originator = config['orig']
$history = config['hist']
$embedbext = config['bext']

#BWF Metaedit Function
def EmbedBEXT(targetfile)
  moddatetime = File.mtime(targetfile)
  moddate = moddatetime.strftime("%Y-%m-%d")
  modtime = moddatetime.strftime("%H:%M:%S")

  #Get Input Name for Description and OriginatorReference
  file_name = File.basename(targetfile)
  originatorreference = File.basename(targetfile, '.wav')
  if originatorreference.length > 32
    originatorreference = "See Description for Identifiers"
  end
  bwfcommand = Bwfmetaeditpath + ' --reject-overwrite ' + '--Description=' + "'" + file_name + "'"  + ' --Originator=' + "'" + $originator + "'" + ' --History=' + "'" + $history + "'" + ' --OriginatorReference=' + "'" + originatorreference + "'" + ' --OriginationDate=' + moddate + ' --OriginationTime=' + modtime + ' --MD5-Embed ' + "'" + targetfile + "'"
  system(bwfcommand)
end

#Function for adjusting buffer
def BufferCheck(sr)
  if sr == '96000'
    $soxbuffer = '50000'
  elsif sr == '48000'
    $soxbuffer = '25000'
  elsif sr == '44100'
    $soxbuffer = '23000'  
  end
end

# GUI App
Shoes.app(title: "AudioRecorder2", width: 600, height: 625) do
  style Shoes::Para, font: "Helvetica"
  background aliceblue
  @logo = image("Resources/audiorecorder_small_1.png", left: 160)
  animate(10) do |frame|
    if frame.to_i.even?
      @logo.path = "Resources/audiorecorder_small_3.png"
    else
      @logo.path = "Resources/audiorecorder_small_2.png"
    end
  end

    def PostRecord(targetfile)
      window(title: "Post-Record Options", width: 600, height: 560) do
        style Shoes::Para, font: "Helvetica"
        background aliceblue
        trimcheck = nil
        @pretrim = $outputdir + '/' + File.basename(targetfile, File.extname(targetfile)) + '_untrimmed' + '.wav'
        @finaloutput = targetfile
        @trimtemp = $outputdir + '/' + 'trim_tempfile.wav'
        stack margin: 15 do
          border gainsboro, strokewidth: 6
          waveform = image($waveform_pic)
        end
        @start_trim = nil
        @end_trim_length = nil

        def SetUpTrim(input)
          ffprobe_command = Ffprobepath + ' -print_format json -show_streams ' + "'" + input + "'"
          $ffprobeout = JSON.parse(`#{ffprobe_command}`)
          @duration_json = $ffprobeout['streams'][0]['duration']
          @duration =@duration_json.to_f
          if ! @end_trim_length.nil? &&  @start_trim_length != "AUTO"
            $end_trim_opt = @duration - @end_trim_length - @start_trim_length
          elsif ! @end_trim_length.nil?
            $end_trim_opt = @duration - @end_trim_length
          end
          if @start_trim_length != "AUTO"
            $start_trim_opt = ' -ss ' + @start_trim_length.to_s
          end
        end
        stack margin: 15 do
          background lightcyan
          para "Press 'Preview' to hear the file you recorded. To trim file, enter the amount (in seconds) to trim from the start and end of the file and press 'Trim'.\n\nIf 'Start Trim' is set to Auto, auto-trim will be applied to start of file.  If no trim at start is desired set this to be empty. After trimming, a preview window will open for your new file. Trim can be run as many times as is neccessary.\n\nPress 'Finish' to quit"
        end
        flow do
          @start_trim_length = "AUTO"
          para 'Start Trim'
          start_trim_input = edit_line text = @start_trim_length do
            if start_trim_input.text.downcase == "auto"
              @start_trim_length = "AUTO"
            else
              @start_trim_length = start_trim_input.text.to_f
            end
          end
        end
        flow do
          para 'End Trim'
          end_trim_input = edit_line do
            @end_trim_length = end_trim_input.text.to_f
          end
        end
        flow do
          preview = button "Preview"
          preview.click do
            if trimcheck.nil?
              command = Mpvpath + ' --force-window --no-terminal --keep-open=yes --title="Preview" --geometry=620x620 -lavfi-complex "[aid1]asplit=3[ao][a][b],[a]showwaves=600x240:n=1[a1],[a1]drawbox=0:0:600:240:t=120[a2],[b]showwaves=600x240:mode=cline:colors=0x00FFFF:split_channels=1[b2],[a2][b2]overlay[vo]" ' + '"' + @finaloutput + '"'
              system(command)
            else
              command = Mpvpath + ' --force-window --no-terminal --keep-open=yes --title="Preview" --geometry=620x620 -lavfi-complex "[aid1]asplit=3[ao][a][b],[a]showwaves=600x240:n=1[a1],[a1]drawbox=0:0:600:240:t=120[a2],[b]showwaves=600x240:mode=cline:colors=0x00FFFF:split_channels=1[b2],[a2][b2]overlay[vo]" ' + '"'+ @pretrim + '"'
              system(command)
            end
          end 
          trim = button "Trim"
          trim.click do
            #set up trim
            if trimcheck.nil?
              SetUpTrim(@finaloutput)
              File.rename(@finaloutput, @pretrim)
              trimcheck = 1
            else
              SetUpTrim(@pretrim)
            end
            if @start_trim_length == "AUTO"
              if ! @end_trim_length.nil? && @end_trim_length != 0.0
                precommand = Ffmpegpath + ' -i ' + '"' + @pretrim + '"' + ' -af silenceremove=start_threshold=-57dB:start_duration=1:start_periods=1 -f wav -c:a ' + $codec_choice  + ' -ar ' + $sample_rate_choice + ' -y -rf64 auto ' + @trimtemp
                system(precommand)
                SetUpTrim(@trimtemp)
                postcommand = Ffmpegpath + " -i #{@trimtemp} -c copy -y -rf64 auto " + ' -t ' + $end_trim_opt.to_s + ' "' + @finaloutput + '"'
                system(postcommand)
                File.delete(@trimtemp)
              else
                command = Ffmpegpath + ' -i ' + '"' + @pretrim + '"' + ' -af silenceremove=start_threshold=-57dB:start_duration=1:start_periods=1 -f wav -c:a ' + $codec_choice  + ' -ar ' + $sample_rate_choice + ' -y -rf64 auto ' + '"' + @finaloutput + '"'
                system(command)
              end
            else
              if ! @end_trim_length.nil? && @end_trim_length != 0.0
                command = Ffmpegpath + ' ' + $start_trim_opt + ' -i ' + '"' + @pretrim + '"' + ' -c copy -y -rf64 auto ' + ' -t ' + $end_trim_opt.to_s + ' "' + @finaloutput + '"'
                system(command)
              else
                command = Ffmpegpath + ' ' + $start_trim_opt + ' -i ' + '"' + @pretrim + '"' + ' -c copy -y -rf64 auto ' + ' "' + @finaloutput + '"'              
                system(command)
              end
            end
            command = Mpvpath + ' --force-window --no-terminal --keep-open=yes --title="Preview" --geometry=620x620 -lavfi-complex "[aid1]asplit=3[ao][a][b],[a]showwaves=600x240:n=1[a1],[a1]drawbox=0:0:600:240:t=120[a2],[b]showwaves=600x240:mode=cline:colors=0x00FFFF:split_channels=1[b2],[a2][b2]overlay[vo]" ' + '"' + @finaloutput + '"'
            system(command)
          end
          close = button "Finish"
          close.click do
            File.delete($waveform_pic)
            close()
          end
        end
      end
    end

  flow margin: 2 do
    para "Select Channel(s)"
    if sox_channels == '1 2'
      sox_channels_saved = '1 and 2'
    elsif sox_channels == '3 4'
      sox_channels_saved = '3 and 4'   
    else
      sox_channels_saved = sox_channels
    end
    channels = list_box items: ["1", "2", "1 and 2", "3 and 4"],
    width: 100, choose: sox_channels_saved do |list|
      if list.text == '1 and 2'
        sox_channels = '1 2'
      elsif list.text == '3 and 4'
        sox_channels = '3 4'
      else
        sox_channels = list.text
      end
      if sox_channels == "1 2"
        ffmpeg_channels = 'stereo'
      elsif sox_channels == "3 4"
        ffmpeg_channels = 'stereo'
      else
        ffmpeg_channels = 'mono'
      end
    end
    para "\n\n"
    para "Sample Rate"
    if $sample_rate_choice == '44100'
      sample_rate_saved = "44.1 kHz"
    elsif $sample_rate_choice == '48000'
      sample_rate_saved = "48 kHz"
    elsif $sample_rate_choice == '96000'
      sample_rate_saved = "96 kHz"
    end  
    samplerate = list_box items: ["44.1 kHz", "48 kHz", "96 kHz"],
    width: 100, choose: sample_rate_saved do |list|
      if list.text == '44.1 kHz'
        $sample_rate_choice = '44100'
      elsif list.text == '48 kHz'
        $sample_rate_choice = '48000'
      elsif list.text == '96 kHz'
        $sample_rate_choice = '96000'
      end
    end
    para "\n\n"
    para "Bit Depth"
    if $codec_choice == 'pcm_s16le'
      codec_saved = "16 bit"
    elsif $codec_choice == 'pcm_s24le'
      codec_saved = "24 bit"  
    end
    bitdepth = list_box items: ["16 bit", "24 bit"],
    width: 100, choose: codec_saved do |list|
      if list.text == '16 bit'
        $codec_choice = 'pcm_s16le'
      elsif list.text == '24 bit'
        $codec_choice = 'pcm_s24le'
      end      
    end
  end

  stack margin: 10 do
    button "Choose Output Directory" do
      $outputdir = ask_open_folder
      @destination.replace "#{$outputdir}"
    end
    flow do
      destination_prompt = para "File will be saved to:"
      @destination = para "#{$outputdir}", underline: "single"
    end 
  end

    stack margin: 10 do
    button "Choose File Name" do
      $filename = ask("Please Enter File Name")
      if $filename == ''
        @outputfile.replace "#{$filename}"
      else
        @outputfile.replace "#{$filename}.wav"
      end
    end
    flow do
      output_prompt = para "File will be saved as:"
      @outputfile = para "#{$filename}", underline: "single"
    end
  end

  flow do
    preview = button "Preview"
    preview.click do
      BufferCheck($sample_rate_choice)
      Soxcommand = Soxpath + ' -r ' + $sample_rate_choice + ' -b 32 -L -e signed-integer --buffer ' + $soxbuffer + ' -p remix ' + sox_channels
      FFmpegSTART = Ffmpegpath + ' -channel_layout ' + ffmpeg_channels + ' -i - '
      FFmpegPreview = '-f wav -c:a ' + 'pcm_s16le -dither_method triangular' + ' -ar ' + '44100' + ' -'
      FFplaycommand = Ffplaypath + ' -window_title "AudioRecorder" -f lavfi ' + '"' + 'amovie=\'pipe\:0\'' + ',' + FILTER_CHAIN + '"' 
      ffmpegcommand = FFmpegSTART + FFmpegPreview
      command = Soxcommand + ' | ' + ffmpegcommand + ' | ' + FFplaycommand
      system(command)
    end

    record = button "Record"
    record.click do
      if ! defined? $record_iteration
        $record_iteration = 1
      else
        $record_iteration = $record_iteration + 1
      end
      if $outputdir.nil?
        $outputdir = ''
      end
      $waveform_pic = $outputdir + '/' + 'AUDIORECORDERTEMP' + $record_iteration.to_s + '.jpg'
      BufferCheck($sample_rate_choice)
      @tempfileoutput = '"' + $outputdir + '/' + $filename + '_temp.wav' + '"'
      @fileoutput = $outputdir + '/' + $filename + '.wav'
      if $filename == ''
        alert "Please enter an output file name!"
      elsif File.exist?(@fileoutput)
        alert "A File named #{$filename} already exists at that location!"
      elsif ! File.exist?($outputdir)
        alert "Please enter a valid output Directory!"
      else
        Soxcommand = Soxpath + ' -r ' + $sample_rate_choice + ' -b 32 -L -e signed-integer --buffer ' + $soxbuffer + ' -p remix ' + sox_channels
        FFmpegSTART = Ffmpegpath + ' -channel_layout ' + ffmpeg_channels + ' -i - '
        FFmpegRECORD = '-f wav -c:a ' + $codec_choice  + ' -dither_method triangular -ar ' + $sample_rate_choice + ' -metadata comment="" -y -rf64 auto ' + @tempfileoutput
        FFmpegPreview = ' -f wav -c:a ' + 'pcm_s16le -dither_method triangular' + ' -ar ' + '44100' + ' -'
        FFplaycommand = Ffplaypath + ' -window_title "AudioRecorder" -f lavfi ' + '"' + 'amovie=\'pipe\:0\'' + ',' + FILTER_CHAIN + '"' 
        ffmpegcommand = FFmpegSTART + FFmpegRECORD + FFmpegPreview
        syscommand1 = Soxcommand + ' | ' + ffmpegcommand + ' | ' + FFplaycommand
        syscommand2 = Ffmpegpath + ' -i ' + @tempfileoutput + ' -lavfi showwavespic=split_channels=1:s=500x150:colors=blue -y ' + $waveform_pic + ' -c copy ' + "'" + @fileoutput + "'" + ' && rm ' + @tempfileoutput
        system(syscommand1) && system(syscommand2)
        if $embedbext == 'true'
          EmbedBEXT(@fileoutput)
        end
        PostRecord(@fileoutput)   
      end
    end

    button "Edit BWF Metadata" do
      window(title: "Edit BWF Metadata", width: 600, height: 500) do
        background aliceblue
        stack do
          para "Please Make Selections"
          para "Originator"
          originator_choice = edit_line text = $originator do
            $originator = originator_choice.text
          end
          para "Coding History"
          history_choice = edit_box text = $history do
            $history = history_choice.text
          end
          flow do
            bextswitch = check
            if $embedbext == 'true'
              bextswitch.checked = true
            elsif $embedbext =='false'
              bextswitch.checked = false
            end 
            para "Embed BFW Metadata?"
            bextswitch.click do
              if bextswitch.checked?
                $embedbext = 'true'
              else
                $embedbext = 'false'
              end
            end
          end
          button "Save Settings" do
            config['destination'] = $outputdir
            config['samplerate'] = $sample_rate_choice
            config['channels'] = sox_channels
            config['codec'] = $codec_choice
            config['orig'] = $originator
            config['hist'] = $history
            config['bext'] = $embedbext
            File.open(configuration_file, 'w') {|f| f.write config.to_yaml }
            close()
          end
          close = button "Cancel"
          close.click do
            close()
          end
        end
      end
    end

    button "Save Settings" do
      config['destination'] = $outputdir
      config['samplerate'] = $sample_rate_choice
      config['channels'] = sox_channels
      config['codec'] = $codec_choice
      config['orig'] = $originator
      config['hist'] = $history
      config['bext'] = $embedbext
      File.open(configuration_file, 'w') {|f| f.write config.to_yaml }
    end

    exit = button "Quit"
    exit.click do
        exit()
    end
  end
end

#Cascadia Now!
